flax/linen/module.py

```diff
@@ -1104,6 +1104,17 @@ def other_fn(instance, ...):
     if method is None:
       method = self.__call__
     method = _get_unbound_fn(method)
+    # Try to detect if user accidentally passed {'params': {'params': ...} or
+    # {'kernel': ...} as `variables`.
+    if 'params' in variables and isinstance(
+        variables['params'], FrozenDict) and 'params' in variables['params']:
+      raise ValueError(
+          'Found nested params in variables dict. '
+          'Pass variables=params instead of `{\'params\':params}`.'
+      )
+    if 'kernel' in variables and 'params' not in variables:
```

Comment:
Why did you add a specific check for `kernel`? Don't we want this error to be raised always if `params` is not in variables?

Your test now passes because `kernel` is in the variable dict, but there are many modules that don't have a `kernel` param, and then we still want to throw an error.

Reply:
I see, had mistakenly thought there was something special about the {"kernel": ...} case mentioned in the task description. 
Do I now understand correctly that {"kernel":...} was mentioned only as an example of the general case where the pytree is missing 'params'?

Reply:
@marcvanzee while updating this conditional I realised that the `{"kernel": ...}` case, where I assume the user had a correctly structured params dict but accidentally passed `variables=variables['params']`, could be tricky to cover properly. That's because I can't simply make apply() enforce that `variables` is a dict that contains a 'params' key, since in our tests we have something like this being used as a legitimate value for `variables`:
```
(Pdb) variables
FrozenDict({
    test: {
        inner: {
            baz: DeviceArray([1.], dtype=float32),
        },
    },
})
``` 
I'm not so familiar with real usecases yet.  Could someone advise on whether it's better to:
A) Enforce the 'params' key constraint, and update all the tests, or
B) Don't go out of our way to check for the {"kernel": ...} case, assuming it's much rarer than `{'params': {'params': ...}}`, or
C) Come up with a better heuristic? If so, any thoughts? 
Thanks!

Reply:
Taking option B for now. Filed #1909 to document findings.